xmppserver/src/main/java/org/jivesoftware/openfire/net/HappyEyeballsResolver.java

```diff
@@ -0,0 +1,324 @@
+package org.jivesoftware.openfire.net;
+
+import net.jcip.annotations.GuardedBy;
+
+import java.net.Inet6Address;
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.net.UnknownHostException;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.*;
+import java.util.concurrent.*;
+import java.util.function.Supplier;
+
+public class HappyEyeballsResolver
+{
+    private final Duration resolutionDelay;
+
+    private final ThreadPoolExecutor executor;
+
+    private final List<DNSUtil.HostAddress> hostAddresses;
+
+    @GuardedBy("this")
+    private final PriorityQueue<IndexedInetAddress> resolvedHosts;
+
+    @GuardedBy("this")
+    private final ConcurrentMap<Integer, Integer> resultCountByIndex = new ConcurrentHashMap<>();
+
+    @GuardedBy("this")
+    private int preferredNextIndex = 0;
+
+    @GuardedBy("this")
+    private boolean preferredNextFamilyIsIpv4;
+
+    public HappyEyeballsResolver(final List<DNSUtil.HostAddress> hostAddresses)
+    {
+        this(hostAddresses, false, Duration.ofMillis(50));
+    }
+
+    public HappyEyeballsResolver(final List<DNSUtil.HostAddress> hostAddresses, final boolean preferIpv4, final Duration resolutionDelay)
+    {
+        executor = (ThreadPoolExecutor) Executors.newFixedThreadPool(hostAddresses.size());
+        this.hostAddresses = hostAddresses;
+        this.resolvedHosts = new PriorityQueue<>(hostAddresses.size()*3, Comparator.comparing(IndexedInetAddress::getIndex));
+        this.preferredNextFamilyIsIpv4 = preferIpv4;
+        this.resolutionDelay = resolutionDelay;
+    }
+
+    public Duration getResolutionDelay()
+    {
+        return resolutionDelay;
+    }
+
+    public synchronized void start() throws ExecutionException, InterruptedException
+    {
+        for (int i = 0; i < hostAddresses.size(); i++)
+        {
+            int index = i;
+            // Happy Eyeballs dictates that first an AAAA and only then A query is sent out for each host. The Java API
+            // that we're using doesn't give us granular control like that: it's requesting both at the same time.
+            final Supplier<Set<IndexedInetAddress>> solve = () -> {
+                final DNSUtil.HostAddress hostAddress = hostAddresses.get(index);
+                try {
+                    return IndexedInetAddress.from(index, InetAddress.getAllByName(hostAddress.getHost()), hostAddress.getPort(), hostAddress.isDirectTLS());
+                } catch (UnknownHostException e) {
+                    throw new RuntimeException(e);
+                }
+            };
+            solve(solve, index);
+        }
+    }
+
+    // Exposed for unit testing.
```

Comment:
I forget if Guava is in your class path, if it is, `@VisibleForTesting` is handy.

Reply:
oooh, that's handy, thanks!